README.md
# s3000_laser

<h2>Description</h2>
Driver for ROS to read the scan data from the device SICK S3000. This driver is based on the original driver developed for Player/Stage by Brian Gerkey, Kasper Stoy, Richard Vaughan, & Andrew Howard.

This driver controls the SICK S3000 safety laser scanner interpreting its data output. The driver is very basic and assumes the S3000 has already been configured to continuously output its measured data on the RS422 data lines.

<h2>Supported Hardware</h2>
This package should work with SICK S3000 safety laser scanners. 

![Image of Sick s3000](http://wiki.ros.org/s3000_laser?action=AttachFile&do=get&target=SICK_S3000.jpg)

<h2>Nodes
S3000 laser driver.

<h3>Published Topics:</h3>
s3000_laser/scan (sensor_msgs/LaserScan)
    Scan data from the laser, containing either range data or reflector detections. 

<h3>Parameters:</h3>
~port (string, default: "/dev/ttyUSB0")
    Serial port. 

~frame_id (string, default: "/laser")
    Relative frame for the publication of the laser measures. 

~baud_rate (int, default: 500000)
    Serial port baudrate (options: 9600, 19200, 38400, 115200, 500000). 

~serial_parity (string, default: "none")
    Serial port parity (options: "even", "odd", "none"). 

~serial_datasize (int, default: 8)
    Character size mask for serial link (options: 5,6,7,8). 

~range_min (float, default: 0m)
    Min valid range (m) for S3000 data produced. 

~range_max (float, default: 40m)
    Max valid range (m) for S3000 data. 


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package s3000_laser
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.1.1 (2016-08-04)
------------------
* Added package to indigo
* fixed array types
* cleanup
* code cleanup
* sicks3000 class cleanup
* minor cleanup
* Merge pull request `#4 <https://github.com/RobotnikAutomation/s3000_laser/issues/4>`_ from RobotnikAutomation/serial-device-code-cleanup
  Serial device code cleanup
* close port in destructor
* cleanup serial device class interface
* remove all unused code for clarity
  cleanup class interface
* Merge pull request `#3 <https://github.com/RobotnikAutomation/s3000_laser/issues/3>`_ from chataign/parse-reflector-data
  Parse reflector data and new parameters
* set queue size to 1
  clean laser scan at each iteration
* allow configuration of the laser scan from ROS params
* parse reflector data
* moved DTOR to header to allow use from other files
* parse reflector data
* make baud rate and topic name configurable. support for 0.5deg resolution
* S30b support added. Minor fixes
* Adding Changelog file
* Adding packages dependencies and removing unused includes
* Modifying package description for Hydro branch
* Adding files for Hydro
* Initial commit
* Contributors: AliquesTomas, Dani Carbonell, Francois Chataigner, RobotnikRoman, Roman Navarro Garcia, francois


src/sicks3000.cc
/*
*  Copyright (c) 2012, Robotnik Automation, SLL
*
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/
/*
 Desc: Driver for the SICK S3000 laser
 Author: Robotnik Automation SLL (based on sicks3000 by Brian Gerkey, Kasper Stoy, Richard Vaughan, & Andrew Howard for Player/Stage)
 Date: 1 Sept 2012

 The sicks3000 driver controls the SICK S 3000 safety laser scanner interpreting its data output.
 The driver is very basic and assumes the S3000 has already been configured to continuously output
 its measured data on the RS422 data lines.
*/

#include <netinet/in.h>  // htons
#include "s3000_laser/sicks3000.h"

// 1 second of data at 500kbaud
#define DEFAULT_RX_BUFFER_SIZE 500*1024/8

/*!	\fn SickS3000::SickS3000()
 * 	\brief Public constructor
*/
SickS3000::SickS3000( std::string port, int baudrate, std::string parity, int datasize )
: serial_( port.c_str(), baudrate, parity.c_str(), datasize )
{
  rx_count = 0;
  // allocate our recieve buffer
  rx_buffer_size = DEFAULT_RX_BUFFER_SIZE;
  rx_buffer = new uint8_t[rx_buffer_size];
  assert(rx_buffer);

  recognisedScanner = false;
}

SickS3000::~SickS3000() 
{
  delete [] rx_buffer;
}

/*!	\fn int SickS3000::Open()
 * 	\brief Open serial port
 * 	\returns -1 Error
 * 	\returns 0 Ok
*/
bool SickS3000::Open()
{
    return serial_.OpenPort();
}

/*!	\fn int SickS3000::Close()
 * 	\brief Closes serial port
 * 	\returns ERROR
 * 	\returns OK
*/
bool SickS3000::Close()
{
    return serial_.ClosePort();
}

////////////////////////////////////////////////////////////////////////////////
// Set up scanner parameters based on number of results per scan
bool SickS3000::SetScannerParams(sensor_msgs::LaserScan& scan, int data_count)
{
    if (data_count == 761) // sicks3000
    {
        float freq_hz = 16.6;
		
        scan.angle_min          = deg_to_rad(-95);
        scan.angle_max          = deg_to_rad(95);
        scan.angle_increment    = deg_to_rad(0.25);
        scan.scan_time          = 1.0 / freq_hz;
        scan.time_increment     = scan.scan_time / data_count;
        scan.range_min          = 0;
        scan.range_max          = 49;   // check ?

        return true;
    }

    if (data_count == 381) // sicks3000 0.5deg resolution
    {
        float freq_hz = 20.0;

        scan.angle_min          = deg_to_rad(-95);
        scan.angle_max          = deg_to_rad(95);
        scan.angle_increment    = deg_to_rad(0.5);
        scan.scan_time          = 1.0 / freq_hz;
        scan.time_increment     = scan.scan_time / data_count;
        scan.range_min          = 0;
        scan.range_max          = 49;   // check ?

        return true;
    }

    if (data_count == 541) // sicks30b
    {
        float freq_hz = 12.7;

        scan.angle_min          = deg_to_rad(-135);
        scan.angle_max          = deg_to_rad(135);
        scan.angle_increment    = deg_to_rad(0.5);
        scan.scan_time          = 1.0 / freq_hz;
        scan.time_increment     = scan.scan_time / data_count;
        scan.range_min          = 0;
        scan.range_max          = 40;   // check ? 30m in datasheet

        return true;
    }
    
    return false;
}

bool SickS3000::ReadLaser( sensor_msgs::LaserScan& scan )
{
	int bytes_read=0;			// Number of received bytes

	// Read controller messages
	if (serial_.ReadPort(read_buffer_, READ_BUFFER_SIZE, bytes_read)==false) 
	{
	    ROS_ERROR("SickS3000::ReadLaser: Error reading port");
	    return false;
    }

	if (rx_count + bytes_read > rx_buffer_size) 
	{
	   ROS_WARN("S3000 Buffer Full");
	   rx_count = 0; // clear buffer
	   return false;
    }

   memcpy(&rx_buffer[rx_count], read_buffer_, bytes_read);
   rx_count += bytes_read;

   bool bValidData=false;
   ProcessLaserData(scan, bValidData);
   
   return bValidData;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

int SickS3000::ProcessLaserData(sensor_msgs::LaserScan& scan, bool& bValidData)
{
  while(rx_count >= 22)
  {
    // find our continuous data header
    unsigned int ii;
    bool found = false;
    for (ii = 0; ii < rx_count - 22; ++ii)
    {
      if (memcmp(&rx_buffer[ii],"\0\0\0\0\0\0",6) == 0)
      {
        memmove(rx_buffer, &rx_buffer[ii], rx_count-ii);
        rx_count -= ii;
        found = true;
        break;
      }
    }

    if (!found)
    {
      memmove(rx_buffer, &rx_buffer[ii], rx_count-ii);
      rx_count -= ii;
      return 0;
    }

    // get relevant bits of the header
    // size includes all data from the data block number
    // through to the end of the packet including the checksum
    unsigned short size = 2*htons(*reinterpret_cast<unsigned short *> (&rx_buffer[6]));
    //ROS_INFO(" size %d   rx_count %d ", size, rx_count);
    if (size > rx_buffer_size - 26)
    {
      ROS_WARN("S3000: Requested Size of data is larger than the buffer size");
      memmove(rx_buffer, &rx_buffer[1], --rx_count);
      return 0;
    }

    // check if we have enough data yet
    if (size > rx_count - 4)
      return 0;


    unsigned short packet_checksum = *reinterpret_cast<unsigned short *> (&rx_buffer[size+2]);
    unsigned short calc_checksum = CreateCRC(&rx_buffer[4], size-2);
    if (packet_checksum != calc_checksum)
    {
      ROS_WARN("S3000: Checksum's dont match, thats bad (data packet size %d)",size);
      memmove(rx_buffer, &rx_buffer[1], --rx_count);
      continue;
    }

    else
    {
      uint8_t * data = &rx_buffer[20];
      if (data[0] != data[1])
      {
        ROS_WARN("S3000: Bad type header bytes dont match");
      }
      else
      {
        if (data[0] == 0xAA)
        {
          ROS_WARN("S3000: We got a I/O data packet we dont know what to do with it\n");
        }
        else if (data[0] == 0xBB)
        {
          int data_count = (size - 22) / 2;
          if (data_count < 0)
          {
            ROS_WARN("S3000: bad data count (%d)", data_count);
            memmove(rx_buffer, &rx_buffer[size+4], rx_count - (size+4));
            rx_count -= (size + 4);
            continue;
          }
          if (!recognisedScanner)
          {
            // Set up parameters based on number of results.
			if(SetScannerParams(scan, data_count)) recognisedScanner=true; 
		  }

	  // Scan data, clear ranges, keep configuration
    	  scan.ranges.clear();
	  scan.intensities.clear();  // not used
          // scan.ranges_count = data_count;
          scan.ranges.resize(data_count);  //
          for (int ii = 0; ii < data_count; ++ii)
          {
            unsigned short Distance_CM = (*reinterpret_cast<unsigned short *> (&data[4 + 2*ii]));
            Distance_CM &= 0x1fff; // remove status bits
            double distance_m = static_cast<double>(Distance_CM)/100.0;
        	scan.ranges[ii] = static_cast<float> (distance_m);
          }

        // CHECK
	  // Return this flag to let the node know that the message is ready to publish
	  bValidData = true;
        }
        else if (data[0] == 0xCC && data[1] == 0xCC )
        {
            size_t read_pos=2; // start reading after 'CCCC' header
            
            uint16_t reflector_count = *reinterpret_cast<uint16_t*>( &data[read_pos] );
            read_pos += sizeof(reflector_count);

            ROS_DEBUG_STREAM( "reflectors=" << reflector_count );
            bValidData = ( reflector_count > 0 );

            uint32_t reflector_data, num_ranges = ( scan.angle_max - scan.angle_min ) / scan.angle_increment;
    
            scan.header.stamp = ros::Time::now();
            scan.ranges.resize( num_ranges );
            std::fill( scan.ranges.begin(), scan.ranges.end(), scan.range_min-1 );
            scan.intensities.clear();  // not used
            
            for ( size_t i=0; i< reflector_count; ++i )
            {
                reflector_data = *reinterpret_cast<uint32_t*>( &data[read_pos] );
                read_pos += sizeof(reflector_data);
                
                int raw_angle = reflector_data & 0xFFFF; // read 16 bits
                int raw_dist = ( reflector_data >> 16 ) & 0x1FFF; // read 13 bits
                ROS_DEBUG("reflector %lu: angle=%.2fdeg distance=%.2fm", i, raw_angle*0.01, raw_dist*0.01 );

                float scan_angle = scan.angle_max - deg_to_rad(raw_angle*0.01);
                int range_idx = ( scan.angle_max - scan_angle ) / scan.angle_increment;

                if ( range_idx >= 0 && range_idx < scan.ranges.size() ) scan.ranges[range_idx] = raw_dist*0.01;
                else ROS_WARN("invalid reflector angle=%.2frad, idx=%d/%lu", scan_angle, range_idx, scan.ranges.size() );
            }
        }
        else
        {
          ROS_WARN("We got an unknown packet\n");
        }

      }
    }

    memmove(rx_buffer, &rx_buffer[size+4], rx_count - (size+4));
    rx_count -= (size + 4);
    continue;
  }
  return 1;
}


static const unsigned short crc_table[256] = 
{
  0x0000, 0x1021, 0x2042, 0x3063, 0x4084, 0x50a5, 0x60c6, 0x70e7,
  0x8108, 0x9129, 0xa14a, 0xb16b, 0xc18c, 0xd1ad, 0xe1ce, 0xf1ef,
  0x1231, 0x0210, 0x3273, 0x2252, 0x52b5, 0x4294, 0x72f7, 0x62d6,
  0x9339, 0x8318, 0xb37b, 0xa35a, 0xd3bd, 0xc39c, 0xf3ff, 0xe3de,
  0x2462, 0x3443, 0x0420, 0x1401, 0x64e6, 0x74c7, 0x44a4, 0x5485,
  0xa56a, 0xb54b, 0x8528, 0x9509, 0xe5ee, 0xf5cf, 0xc5ac, 0xd58d,
  0x3653, 0x2672, 0x1611, 0x0630, 0x76d7, 0x66f6, 0x5695, 0x46b4,
  0xb75b, 0xa77a, 0x9719, 0x8738, 0xf7df, 0xe7fe, 0xd79d, 0xc7bc,
  0x48c4, 0x58e5, 0x6886, 0x78a7, 0x0840, 0x1861, 0x2802, 0x3823,
  0xc9cc, 0xd9ed, 0xe98e, 0xf9af, 0x8948, 0x9969, 0xa90a, 0xb92b,
  0x5af5, 0x4ad4, 0x7ab7, 0x6a96, 0x1a71, 0x0a50, 0x3a33, 0x2a12,
  0xdbfd, 0xcbdc, 0xfbbf, 0xeb9e, 0x9b79, 0x8b58, 0xbb3b, 0xab1a,
  0x6ca6, 0x7c87, 0x4ce4, 0x5cc5, 0x2c22, 0x3c03, 0x0c60, 0x1c41,
  0xedae, 0xfd8f, 0xcdec, 0xddcd, 0xad2a, 0xbd0b, 0x8d68, 0x9d49,
  0x7e97, 0x6eb6, 0x5ed5, 0x4ef4, 0x3e13, 0x2e32, 0x1e51, 0x0e70,
  0xff9f, 0xefbe, 0xdfdd, 0xcffc, 0xbf1b, 0xaf3a, 0x9f59, 0x8f78,
  0x9188, 0x81a9, 0xb1ca, 0xa1eb, 0xd10c, 0xc12d, 0xf14e, 0xe16f,
  0x1080, 0x00a1, 0x30c2, 0x20e3, 0x5004, 0x4025, 0x7046, 0x6067,
  0x83b9, 0x9398, 0xa3fb, 0xb3da, 0xc33d, 0xd31c, 0xe37f, 0xf35e,
  0x02b1, 0x1290, 0x22f3, 0x32d2, 0x4235, 0x5214, 0x6277, 0x7256,
  0xb5ea, 0xa5cb, 0x95a8, 0x8589, 0xf56e, 0xe54f, 0xd52c, 0xc50d,
  0x34e2, 0x24c3, 0x14a0, 0x0481, 0x7466, 0x6447, 0x5424, 0x4405,
  0xa7db, 0xb7fa, 0x8799, 0x97b8, 0xe75f, 0xf77e, 0xc71d, 0xd73c,
  0x26d3, 0x36f2, 0x0691, 0x16b0, 0x6657, 0x7676, 0x4615, 0x5634,
  0xd94c, 0xc96d, 0xf90e, 0xe92f, 0x99c8, 0x89e9, 0xb98a, 0xa9ab,
  0x5844, 0x4865, 0x7806, 0x6827, 0x18c0, 0x08e1, 0x3882, 0x28a3,
  0xcb7d, 0xdb5c, 0xeb3f, 0xfb1e, 0x8bf9, 0x9bd8, 0xabbb, 0xbb9a,
  0x4a75, 0x5a54, 0x6a37, 0x7a16, 0x0af1, 0x1ad0, 0x2ab3, 0x3a92,
  0xfd2e, 0xed0f, 0xdd6c, 0xcd4d, 0xbdaa, 0xad8b, 0x9de8, 0x8dc9,
  0x7c26, 0x6c07, 0x5c64, 0x4c45, 0x3ca2, 0x2c83, 0x1ce0, 0x0cc1,
  0xef1f, 0xff3e, 0xcf5d, 0xdf7c, 0xaf9b, 0xbfba, 0x8fd9, 0x9ff8,
  0x6e17, 0x7e36, 0x4e55, 0x5e74, 0x2e93, 0x3eb2, 0x0ed1, 0x1ef0
};

unsigned short SickS3000::CreateCRC(const uint8_t *Data, ssize_t length)
{
  unsigned short CRC_16 = 0xFFFF;
  unsigned short i;
  for (i = 0; i < length; i++)
  {
    CRC_16 = (CRC_16 << 8) ^ (crc_table[(CRC_16 >> 8) ^ (Data[i])]);
  }
  return CRC_16;
}


src/SerialDevice.cc
/*
*  Copyright (c) 2012, Robotnik Automation, SLL
* 
*   This file is part of sick-s3000-ros-pkg.
*
*   sick-s3000-ros-pkg is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/
/** \file SerialDevice.cc
 * \author Robotnik Automation S.L.L.
 * \version 1.1
 * \date    2009
 *
 * \brief SerialDevice Class 
 * (C) 2009 Robotnik Automation, SLL
 * Class to manage a serial connection
*/

#include <s3000_laser/SerialDevice.h>
#include <time.h>
#include <string.h> 
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>  /* UNIX standard function definitions */
#include <sys/stat.h>
#include <termios.h>
#include <sys/ioctl.h>
#include <iostream>
#include <errno.h>   /* Error number definitions */
#include <stropts.h>
#include <ros/ros.h>

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

SerialDevice::SerialDevice( const std::string& port, int baudrate, const std::string& parity, int datasize)
: port_( port )
, parity_( parity )
, baudrate_( baudrate )
, datasize_( datasize )
{		
	ROS_INFO_STREAM( "SerialDevice: " << port 
	    << " Parity= " << parity 
	    << " DataSize=" << datasize 
	    << " BaudRate=" << baudrate );
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

SerialDevice::~SerialDevice() 
{
    ClosePort();
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

bool SerialDevice::OpenPort()
{
	serial_port_ = open( port_.c_str(),  O_RDWR | O_NOCTTY | O_NDELAY | O_NONBLOCK );
	
	if( serial_port_ == -1 )
	{
		ROS_WARN( "Error opening serial port=%s", port_.c_str() );
		return false;  // invalid port file	
	}
	
	// set up comm flags
	struct termios comms_flags;
	memset( &comms_flags, 0, sizeof(termios) );
	
	switch(datasize_)
	{
		case 5:
			comms_flags.c_cflag = CS5 | CREAD;
		    break;
		case 6:
			comms_flags.c_cflag = CS6 | CREAD;
		    break;
		case 7:
			comms_flags.c_cflag = CS7 | CREAD;
		    break;
		case 8:
			comms_flags.c_cflag = CS8 | CREAD;
		    break;
		default:
			ROS_WARN( "unsupported datasize=%d", datasize_ );
		    return false;
	}
	
	comms_flags.c_iflag = INPCK;
	comms_flags.c_oflag = 0;
	comms_flags.c_lflag = 0;
	
	if( parity_ == "even" )
	{
		comms_flags.c_cflag |= PARENB;
	}
	else if( parity_ == "odd" )
	{
		comms_flags.c_cflag |= PARODD;
	}
	else if( parity_ == "none" )
	{
		comms_flags.c_cflag &= ~PARODD;
		comms_flags.c_cflag &= ~PARENB;
	}
	
	comms_flags.c_lflag &= ~ICANON;	// TEST
	
	tcsetattr( serial_port_, TCSANOW, &comms_flags );
	tcflush( serial_port_, TCIOFLUSH );
		
	if ( SetTermSpeed(baudrate_) == false )
	    return false;
	
	// Make sure queue is empty
	tcflush( serial_port_, TCIOFLUSH );
	return true;
	
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

bool SerialDevice::ClosePort()
{
	return close(serial_port_) == 0;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

bool SerialDevice::ReadPort( char *buffer, int bytes_to_read, int &bytes_read ) 
{
	bytes_read = read( serial_port_, buffer, bytes_to_read );

    if( bytes_read >= 0 ) 
    {
        return true;
    }
    else if ( errno == EAGAIN ) 
    {
        return true; // nothing to read
    }
    else
    {
        ROS_WARN("SERIAL read error %d %s", errno, strerror(errno) );
        return false;
    }
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

bool SerialDevice::SetTermSpeed(int baudrate)
{
	int baudrate_flag;
	
	switch(baudrate)
	{
		case 9600:
			baudrate_flag = B9600;
			break;
		case 19200:
			baudrate_flag = B19200;
			break;
		case 38400:
			baudrate_flag = B38400;
			break;
		case 115200:
			baudrate_flag = B115200;
			break;
		case 500000:
			baudrate_flag = B500000;
			break;
		default:
		    ROS_WARN("unsupported baudrate=%d", baudrate );
			return false;
	}
	
	struct termios comms_flags;

	if( tcgetattr( serial_port_, &comms_flags ) < 0 )
		return false;
	
	if( cfsetispeed( &comms_flags, baudrate_flag ) < 0 || cfsetospeed( &comms_flags, baudrate_flag ) < 0)
		return false;
	
	if( tcsetattr( serial_port_, TCSAFLUSH, &comms_flags ) < 0 )
		return false;

	return true;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////



include/s3000_laser/sicks3000.h
/*
*  Copyright (c) 2012, Robotnik Automation, SLL
*
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/
/*
 Desc: Driver for the SICK S3000 laser
 Author: Robotnik Automation SLL (based on sicks3000 by Brian Gerkey, Kasper Stoy, Richard Vaughan, & Andrew Howard for Player/Stage)
 Date: 1 Sept 2012

 The sicks3000 driver controls the SICK S 3000 safety laser scanner interpreting its data output.
 The driver is very basic and assumes the S3000 has already been configured to continuously output
 its measured data on the RS422 data lines.
*/
#include <assert.h>
#include <math.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <s3000_laser/SerialDevice.h>

//! Converts degrees to radians
inline double deg_to_rad(double val) { return val*M_PI/180.0; }

// The laser device class.
class SickS3000
{
  public:

    // Constructor
    SickS3000( std::string port, int baudrate, std::string parity, int datasize );

    // Destructor
    ~SickS3000();

    //! Open the port
    bool Open();

    //! Close the port
    bool Close();

    //! Read and process data
    bool ReadLaser( sensor_msgs::LaserScan& scan );
  
  private:

    // Process range data from laser
    // int ProcessLaserData();
    int ProcessLaserData( sensor_msgs::LaserScan& scan_msg, bool& bValidData ); // public periodic function

    // Calculates CRC for a telegram
    static unsigned short CreateCRC(const uint8_t *data, ssize_t len);

    static bool SetScannerParams(sensor_msgs::LaserScan& scan, int data_count);

  protected:

    // serial port
    SerialDevice serial_;

    bool recognisedScanner;

    // rx buffer
    uint8_t* rx_buffer;
    unsigned int rx_buffer_size;
    unsigned int rx_count;
    
    static const size_t READ_BUFFER_SIZE=2000;
    char read_buffer_[READ_BUFFER_SIZE];
};




include/s3000_laser/SerialDevice.h
/*
*  Copyright (c) 2012, Robotnik Automation, SLL
* 
*   This file is part of sick-s3000-ros-pkg.
*
*   sick-s3000-ros-pkg is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/

/** \file SerialDevice.h
 * \author Robotnik Automation S.L.L.
 * \version 1.1
 * \date    2009
 *
 * \brief SerialDevice Class 
 * (C) 2009 Robotnik Automation, SLL
 * Class to manage a serial port connection
*/

#ifndef __SERIALDEV_H
#define __SERIALDEV_H
	
#include <stdio.h>
#include <string>

class SerialDevice
{
private:	
	//Device's name
	const std::string port_;	
	//Parity for input and output: EVEN, ODD, NONE
	const std::string parity_;	
	//BaudRate: 9600, 19200, 38400, 115200, 500000
	const int baudrate_;
	//Character size mask. Values are CS5, CS6, CS7, or CS8.
	const int datasize_;

	int serial_port_; // File descriptor

public:
	SerialDevice( const std::string& port, int baudrate, const std::string& parity, int datasize );	
	virtual ~SerialDevice();
	
	bool OpenPort();
	bool ClosePort();
	bool ReadPort( char *result, int bytes_to_read, int &bytes_read );

private:
	//!	Set serial communication speed.
	bool SetTermSpeed(int speed);

};

#endif


src/s3000_laser.cc
/*********************************************************************
*
*  This program uses the sicks3000  component to get laser scans, and then
*  publishes them as ROS messages
*
*  Copyright (c) 2012, Robotnik Automation, SLL
*
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*********************************************************************/

#include <assert.h>
#include <math.h>
#include <iostream>
#include <boost/format.hpp>

#include "s3000_laser/sicks3000.h"   // s3000 driver from player (Toby Collet / Andrew Howard)
#include "ros/time.h"
#include "self_test/self_test.h"
#include "diagnostic_msgs/DiagnosticStatus.h"
#include "diagnostic_updater/diagnostic_updater.h"
#include "diagnostic_updater/update_functions.h"
#include "diagnostic_updater/DiagnosticStatusWrapper.h"

#include "sensor_msgs/LaserScan.h"

#include "tf/transform_broadcaster.h"
#include "s3000_laser/enable_disable.h"

using namespace std;

class s3000node {

private:
	SickS3000 *laser; 
	sensor_msgs::LaserScan reading;

	string port;
	int baud_rate;

	self_test::TestRunner self_test_;
	diagnostic_updater::Updater diagnostic_;

	ros::NodeHandle node_handle_;
	ros::NodeHandle private_node_handle_;
	ros::Publisher laser_data_pub_;

	//! publish its transform to base_link
	bool publish_tf_;
	tf::TransformBroadcaster laser_broadcaster;
	//! Flag to enable/disable the scan publication
	bool publish_scan_;	

	bool running;

	int error_count_;
	int slow_count_;
	std::string was_slow_;
	std::string error_status_;
	//! Name of the frame associated
	string frame_id_;
	string topic_name;
	double desired_freq_;
	diagnostic_updater::FrequencyStatus freq_diag_;
	//! Enables/disables the scan publication
	ros::ServiceServer enable_disable_srv_;

public:
	
	s3000node(ros::NodeHandle h) : self_test_(), diagnostic_(), 
	node_handle_(h), private_node_handle_("~"), 
	error_count_(0), 
	slow_count_(0), 
	desired_freq_(20), 
	freq_diag_(diagnostic_updater::FrequencyStatusParam(&desired_freq_, &desired_freq_, 0.05))
	{
	    int serial_datasize;
	    string serial_parity;
	    float angle_min_deg, angle_max_deg, angle_increment_deg;
	    
		ros::NodeHandle laser_node_handle(node_handle_, "s3000_laser");
		private_node_handle_.param("port", port, string("/dev/ttyUSB0"));
		private_node_handle_.param("baud_rate", baud_rate, 500000);
		private_node_handle_.param<float>("range_min", reading.range_min, 0 );
		private_node_handle_.param<float>("range_max", reading.range_max, 40 );
		private_node_handle_.param<float>("angle_min", angle_min_deg, -100 );
		private_node_handle_.param<float>("angle_max", angle_max_deg, 100 );
		private_node_handle_.param<float>("angle_increment", angle_increment_deg, 0.25 );
		private_node_handle_.param("topic_name", topic_name, string("/scan"));
		private_node_handle_.param<bool> ("publish_tf", publish_tf_, false);
		private_node_handle_.param<bool> ("publish_scan", publish_scan_, true);
		private_node_handle_.param<int> ("serial_datasize", serial_datasize, 8);
		private_node_handle_.param<string> ("serial_parity", serial_parity, "none");
		laser_data_pub_ = laser_node_handle.advertise<sensor_msgs::LaserScan>(topic_name, 1);
		running = false;
		private_node_handle_.param("frame_id", frame_id_, string("/laser"));
		
		reading.angle_min = deg_to_rad(angle_min_deg);
		reading.angle_max = deg_to_rad(angle_max_deg);
		reading.angle_increment = deg_to_rad(angle_increment_deg);
		reading.header.frame_id = frame_id_;
			
		self_test_.add("Connect Test", this, &s3000node::ConnectTest);

		diagnostic_.add( freq_diag_ );
		diagnostic_.add( "Laser S3000 Status", this, &s3000node::deviceStatus );
		ROS_INFO("s3000node: Port = %s", port.c_str());
		ROS_INFO("s3000node: Baudrate = %d", baud_rate);
		
		// Advertises new service to enable/disable the scan publication
		enable_disable_srv_ = private_node_handle_.advertiseService("enable_disable",  &s3000node::EnableDisable, this);
		
		// Create SickS3000 in the given port
		laser = new SickS3000( port, baud_rate, serial_parity, serial_datasize );
	}

	~s3000node()
	{
		stop();
	}
	
	bool spin(){

		// sensor_msgs::LaserScan data;
		ros::Rate r(desired_freq_);

		while (!ros::isShuttingDown()) // Using ros::isShuttingDown to avoid restarting the node during a shutdown.
		{
			if (start() == 0)
			{
				while(node_handle_.ok()) {
					getData( reading );           

					self_test_.checkTest();
					diagnostic_.update();
					ros::spinOnce();
					r.sleep();          
				}
			} else {
				// No need for diagnostic here since a broadcast occurs in start
				// when there is an error.
				usleep(1000000);
				self_test_.checkTest();
				ros::spinOnce();
			}
		}

		stop();

		return true;
	}

private:

	int start()
	{
		stop();
		laser->Open();
		diagnostic_.setHardwareID("Laser Ranger");
		ROS_INFO("Laser Ranger sensor initialized.");
		freq_diag_.clear();
		running = true;
		return(0);
	}


	int stop()
	{
		if(running)
		{
			laser->Close();
			running = false;
		}

		return(0);
	}


	void ConnectTest(diagnostic_updater::DiagnosticStatusWrapper& status)
	{
		laser->Open();
		status.summary(0, "Connected successfully.");
	}
	
	//! Gets the sensor data and publishes 
	void getData(sensor_msgs::LaserScan& scan)
	{
		//// If valid scan, publish it
		if ( laser->ReadLaser(scan) ) 
		{
			scan.header.stamp = ros::Time::now();
			scan.header.frame_id = frame_id_;
			if(publish_scan_)	// Publishes only if enabled
				laser_data_pub_.publish( scan );
			freq_diag_.tick();
		}
	}

	//! Updates the diagnostic status
	void deviceStatus(diagnostic_updater::DiagnosticStatusWrapper &status)
	{
		if (!running)
			status.summary(2, "Laser Ranger is stopped");
		else if (!was_slow_.empty())
		{
			status.summary(1, "Excessive delay");
			was_slow_.clear();
		}
		else{
			
			status.summary(0, "Laser Ranger is running");
			
		}
		status.add("Device", port);
		status.add("TF frame", frame_id_);
		status.add("Error count", error_count_);
		status.add("Excessive delay", slow_count_);
		status.add("Scan publication", publish_scan_);
	}

    //! Service callback to enable/disable scan publication
	bool EnableDisable(s3000_laser::enable_disable::Request &req, s3000_laser::enable_disable::Response &res ){
		publish_scan_ = req.value;

		ROS_INFO("s3000_node::EnablaDisable: Setting scan publish to %d", req.value);
		res.ret = true;
		
		return true;
	}
};


int main(int argc, char** argv)
{
	ros::init(argc, argv, "s3000_node");

	ros::NodeHandle nh;

	s3000node s3000n(nh);
	s3000n.spin();

	return(0);
}
